jina/peapods/grpc/__init__.py

```diff
@@ -0,0 +1,148 @@
+                raise ex
+
+    @staticmethod
+    def send_ctrl_msg(pod_address: str, command: str):
+        """
+        Sends a control message via gRPC to pod_address
+        :param pod_address: the pod to send the command to
+        :param command: the command to send (TERMINATE/ACTIVATE/...)
+        :returns: Empty protobuf struct
+        """
+        stub = Grpclet._create_grpc_stub(pod_address, is_async=False)
+        response = stub.Call(ControlMessage(command))
+        return response
+
+    @staticmethod
+    def _create_grpc_stub(pod_address, is_async=True):
+        if is_async:
+            channel = grpc.aio.insecure_channel(
+                pod_address,
+                options=[
+                    ('grpc.max_send_message_length', -1),
+                    ('grpc.max_receive_message_length', -1),
+                ],
+            )
+        else:
+            channel = grpc.insecure_channel(
+                pod_address,
+                options=[
+                    ('grpc.max_send_message_length', -1),
+                    ('grpc.max_receive_message_length', -1),
+                ],
+            )
+
+        stub = jina_pb2_grpc.JinaDataRequestRPCStub(channel)
+
+        return stub
+
+    async def _add_envelope(self, msg, routing_table):
+        new_envelope = jina_pb2.EnvelopeProto()
+        new_envelope.CopyFrom(msg.envelope)
+        new_envelope.routing_table.CopyFrom(routing_table.proto)
+        new_message = Message(request=msg.request, envelope=new_envelope)
+
+        return new_message
+
+    async def close(self, *args, **kwargs):
+        """Stop the Grpc server
+        :param args: Extra positional arguments
+        :param kwargs: Extra key-value arguments
+        """
+        self._logger.debug('Close grpc server')
+        await self._grpc_server.stop(grace=True)
+
+    async def start(self):
+        """
+        Starts this Grpclet by starting its gRPC server
+        """
+        self._grpc_server = grpc.aio.server(
+            options=[
+                ('grpc.max_send_message_length', -1),
+                ('grpc.max_receive_message_length', -1),
+            ]
+        )
+
+        jina_pb2_grpc.add_JinaDataRequestRPCServicer_to_server(self, self._grpc_server)
+        bind_addr = f'{self.args.host}:{self.args.port_in}'
+        self._grpc_server.add_insecure_port(bind_addr)
+        self._logger.debug(f'Binding gRPC server for data requests to {bind_addr}')
+        await self._grpc_server.start()
+        await self._grpc_server.wait_for_termination()
+
+    async def Call(self, msg, *args):
+        """Processes messages received by the GRPC server
+        :param msg: The received message
+        :param args: Extra positional arguments
+        :return: Empty protobuf struct, necessary to return for protobuf Empty
+        """
+        if self.callback:
+            if inspect.iscoroutinefunction(self.callback):
+                self._loop.create_task(self.callback(msg))
```

Comment:
why can we not do `asyncio.create_task(...)` or `await callback` directly if we need to wait for completion. Do we need the loop object?

Reply:
I will use `asyncio.create_task()`.

I am not awaiting the result because I actually don't care for it and want to respond to the gRPC call right away. For backpressure/bottleneck detection blocking here might be useful but I am not sure we should do that here now. Maybe later?